### PR TITLE
ICM20602 and ICM20608g: new standalone driver (8 kHz gyro, 4 kHz accel)

### PR DESCRIPTION
**Pull request by dagar**
This is a new driver for the Invensense ICM20602 that uses the sensor's FIFO and SPI DMA (on px4_fmu-v4) to transfer raw gyro data at 8 kHz and accel at 4 kHz. It can also do gyro at 32 kHz, but at that point the bus utilization is a bit too high to peacefully coexist with other sensors.

The intent is to keep these opinionated drivers simple and optimized to the point where it'll likely be easier to carry separate drivers for each sensor with minor differences rather than trying to support everything in one place (see current mpu6000 mess). We'll see how this actually pans out for the ICM20608, ICM20689, etc.

Originally I was going to do a pass and enable SPI DMA everywhere first (https://github.com/PX4/Firmware/pull/12436), but given various problems and delays in that area I think we can move forward incrementally like this.

Features
 - 8 kHz gyro (32 kHz possible)
 - 4 kHz accel
 - FIFO emptied at 1 kHz (we can consider doing this faster or slower as appropriate)
 - clip counters
 - new low pass filter that operates on an array of data (LowPassFilter2pArray)
 - new sensor messages for better visibility
    - sensor_{accel, gyro}_fifo: full raw data for optional logging and analysis
    - sensor_{accel, gyro}_status: metadata, clipping, etc
    - sensor_{accel, gyro}_integrated: delta angles and delta velocities 
    - eventually these will replace the existing sensor_{accel, gyro} monolith

#### IMU Signal Processing
Failure to detect vibration induced inertial navigation failures early enough during system integration or operation has contributed to a number of vehicle loss incidents. In the majority of cases this has been caused by structural resonances excited by motor and propeller blade bassage frequency vibration.
The current practice of using the IMU’s internal DLPF is  a major blocker for the following reasons:
 1. IMU internal sampling is then limited to 1kHz which makes the ADC step susceptible to aliasing on airframes with high RPM/geared motors and high blade count propellers
 2. The use of the IMU’s internal DLPF means that high frequency vibration clipping of the ADC cannot be detected externally until it has affected the INS
 3. Data cannot be logged at the ADC sampling frequency so it is not possible to use a data driven approach to resolving vibration induced navigation failures.

Experience working with companies using PX4 has shown that the majority do not understand how propulsion system forcing, structural dynamics, IMU sampling and vibration isolation mount design interact to affect navigation system performance and for those that do, the inability to log the required data prevents the use of normal analytical methods.

The signal chain for autopilot feedback and navigation needs to be split for the following reasons:
 1. Logging of raw unfiltered data from the sensors ADC is required if an analytical approach to resolving vibration induced navigation failures is to be adopted.
 2. Angular rate control feedback requires data with minimum phase low/group delay with a maximum acceptable level of high frequency content that is application specific. 
 3. The best general purpose filter in terms of high frequency noise rejection for a given phase delay measured at the rate controller gain cross-over frequency of 5-15Hz is an IIR second order bi-quad filter with a Q of 1/sqrt(2).
 4. The angular rate controller should be synchronised to the imu to reduce inter-sampling delay.
 5. The angular rate controller should run at a higher rate than the attitude and position/height controllers and therefore needs to run at a faster rate than the INS/EKF. 
 6. The angular rate controller may be required to run at 1kHz or higher for some applications.
 7. The INS/EKF uses delta angles and delta velocities that are obtained by integrating high rate gyro and accel measurements in body frame across the INS/EKF update interval.
 8. The INS/EKF is very sensitive to errors caused by clipping and aliasing. 
 9. Correction for coning errors introduced when downsampling needs to applied as part of the downsampling integration algorithm.
10. It is unlikely that INS data would be required faster than 250Hz.

image

### CODE REVIEW

**Comment by LorenzMeier** [comment]
Awesome!

**Comment by dagar** [comment]
FYI @priseborough @RomanBapst

File: src/systemcmds/top/CMakeLists.txt [hunk 1]
```diff
@@ -33,7 +33,7 @@
 px4_add_module(
 	MODULE systemcmds__top
 	MAIN top
-	PRIORITY "SCHED_PRIORITY_MAX"
+	PRIORITY "SCHED_PRIORITY_DEFAULT"
```
Comment (bkueng) [inline src/systemcmds/top/CMakeLists.txt]: This was set to max on purpose. I had several cases where this helped.

**Comment by BazookaJoe1900** [comment]
That sound huge potential for improvement of the flight control! cool.
I there a theoretical/practical knowledge how its really going to improve the flight control?
Is it going to allow higher vibration tolerance?

File: src/systemcmds/top/CMakeLists.txt (hunk 1 above)
Comment (dagar) [reply src/systemcmds/top/CMakeLists.txt]: I was going to follow up with you. At SCHED_PRIORITY_MAX top can interfere with driver scheduling enough to cause an occasional fifo overflow and reset. I'd like to see if we can find a line in between.

Comment (dagar) [reply src/systemcmds/top/CMakeLists.txt]: How about an argument to optionally run at max priority, but by default at least lower than the sensor bus WQ threads?

File: msg/sensor_accel_control.msg [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+# WARNING: Not recommend for custom development.
+#  This topic is part of an incremental refactoring of the sensor pipeline and will likely disappear post PX4 release v1.10
+
+uint32 device_id	# unique device ID for the sensor that does not change between power cycles
+
+uint64 timestamp	# time since system start (microseconds)
+
+uint64 timestamp_sample	# time the raw data was sampled (microseconds)
+
+float32[3] xyz		# filtered angular velocity in the board axis in rad/s
```
[the remark this answers is not in the document]
Comment (dagar) [reply msg/sensor_accel_control.msg]: Nice catch. This message actually isn't even used yet, but will be soon. Thanks

[the remark this answers is not in the document]
Comment (dagar) [reply msg/sensor_accel_control.msg]: Fixed.

File: msg/sensor_accel_fifo.msg [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+uint64 timestamp_sample	# time since system start (microseconds)
+
+float32 dt		# delta time between samples (microseconds)
+float32 scale
+
+uint8 samples		# number of valid samples
+
+int16[8] x		# angular velocity in the NED X board axis in rad/s
+int16[8] y		# angular velocity in the NED Y board axis in rad/s
+int16[8] z		# angular velocity in the NED Z board axis in rad/s
```
[the remark this answers is not in the document]
Comment (dagar) [reply msg/sensor_accel_fifo.msg]: Fixed.

**Comment by dagar** [comment]
Updated with some background information (thanks to @priseborough).

File: src/systemcmds/top/CMakeLists.txt (hunk 1 above)
Comment (dagar) [reply src/systemcmds/top/CMakeLists.txt]: At the time top was bumped to SCHED_PRIORITY_MAX the IMU drivers and a few other things were running out of ISRs (hrt_call_every).

Comment (dagar) [reply src/systemcmds/top/CMakeLists.txt]: Continued in https://github.com/PX4/Firmware/pull/13115.

Comment (bkueng) [reply src/systemcmds/top/CMakeLists.txt]: Out of curiosity: is a fifo overflow handled gracefully?

**Comment by priseborough** [comment]
I did a quick test of  this branch with sensor comparison logging on an earlier version of the pixracer with a icm20608g IMU.

The  test involved shaking the board vertically with a low amplitude  acceleration at three frequency points. The  first was a 900 -> 1100 Hz, followed by 3900 => 4100 and then 7900 ->8100.

The following shows the result obtained with the legacy code (log here: https://review.px4.io/plot_app?log=1187443a-c497-4966-93ab-60d99c60ee81)

Both IMU's aliased in both the 1kHz. and 4kHz band. The first third is the 1kHz band testing, the middle third is 4kHz and the last third the 8kHz. There was minimal aliasing at 8kHz.

before

Repeating the test using the code from this PR (log here https://logs.px4.io/plot_app?log=176fec85-661c-441f-bdf5-07a41811b4e1) shows  that IMU0 now no longer aliases at 1kHz, eg

after

Both continue to alias at 4kHz as expected. The accelerometers on both IMU's lack effective internal anti-aliasing protection, highlighting why a high sampling rate and mechanical vibration isolation is desirable.

**Comment by priseborough** [comment]
BTW, the vibration test setup is a small consumer grade vibration speaker using a mobile phone signal generator app.

Screen Shot 2019-10-18 at 4 23 13 pm

**Comment by dagar** [comment]
@PX4/testflights could you please give this an initial test flight on a pixracer? Please enable "sensor comparison" (SDLOG_PROFILE).

**Comment by jorge789** [comment]
**Tested on PixRacer V4:**
**Modes Tested**

Position Mode: Good.
Altitude Mode: Good.
Stabilized Mode: Good.
Mission Plan Mode (Automated): Good.
RTL: Good.

**- Procedure**
Arm and Take off in position mode, after flying for approximately one minute, switched to altitude then stabilized mode proceed to switch to mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, once completed all waypoin activate RTL and see landing behaviour.

**Notes:**
**When the vehicle was landing, and I touched the ground, it began to make a few small leaps but did not land completely after a few seconds, the vehicle rose a meter and it was necessary to land manually,the vehicle failed to disarm, to had disarm it manually.**

**Log:**
https://review.px4.io/plot_app?log=90bbbcce-326c-4986-9a1c-718699d6c231

**Tested on PixRacer V4:**

**Modes Tested**
Position mode.

**Note-**He was flying in position when suddenly the car pulled back and began to lose height without control and crashed.

**Log:**
https://review.px4.io/plot_app?log=3aa11e66-7dd6-42d8-83c6-3640f901ccce

**Comment by dagar** [comment]
Thanks @jorge789,

Your log uncovered a bug in the sensors module where failover is incorrectly going back to the primary sensor.

Secondly, the accelerometer in that vehicle is clipping in the z axis. It stops in the last half of the log once you settle into calmer mission flight. 

image

**Comment by dusan19** [comment]
> Your log uncovered a bug in the sensors module where failover is incorrectly going back to the primary sensor.

When the sensor failover happens, sensors module sets the priority of the failed sensor to min priority of 1, which should ensure that that sensor doesn't get voted for again.

It might be that the main sensor failed, got priority 1, and the second sensor got voted for. Then the second sensor failed, got priority 1, and now both sensors had priority 1 which made them equally likely to be voted for, although I don't see why the second sensor failed (the logging rate is too low to see timeouts). The main sensor might have failed due to the high error density.

Is it possible that `_integrator.put(timestamp, val_calibrated, integrated_value, integral_dt)` returns false and the data is not published on the sensor_accel topic thus the validator reports a timeout?

**Comment by dagar** [comment]
@PX4/testflights could you give this another test flight? This time both the pixracer and pixhawk 3 pro. Thanks!

If you want a better view into the vibration of the particular vehicle you should enable this fields in the QGC widget setup (attitude indicator gear icon).

image

**Comment by jorge789** [comment]
**Tested on PixRacer V4:**
**Modes Tested**

Position Mode: Good.
Altitude Mode: Good.
Stabilized Mode: Good.
Mission Plan Mode (Automated): Good.
RTL: Good.

**- Procedure**
Arm and Take off in position mode, after flying for approximately one minute, switched to altitude then stabilized mode proceed to switch to mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, once completed all waypoin activate RTL and see landing behaviour.

**Notes:**
No issues noted, good flight in general. 

**Log:**  https://review.px4.io/plot_app?log=44f75460-10ed-4b3a-aaeb-1f3eecc5d0a9

Vibration

**Comment by jorge789** [comment]
### Tested on PixRacer V4:
Modes Tested

Position Mode: Good.
Altitude Mode: See comments below: 
Stabilized Mode: Good.
Mission Plan Mode (Automated): Good.
RTL: Good.
Land. See comments below: 

**Procedure:** 
Arm and Takeoff in position mode, after flying for approximately one minute, switched to altitude then stabilized mode proceeds to switch to mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, once completed all waypoint activate RTL.

**Notes:
Please note on altitude mode, when doing pitch and the sticks centered the vehicle would climb altitude. On landing it did a couple  small leaps then landed and disarmed. Other than that good flight no visible observations of any issues.** 

**Log:**  https://review.px4.io/plot_app?log=0511e98f-b0d4-456f-b79a-181676f7a9a9

### Tested on Pixhawk Pro V4:
Modes Tested

Position Mode: Good.
Altitude Mode: See comments below: 
Stabilized Mode: Good.
Mission Plan Mode (Automated): Good.
RTL: Good.
Land. See comments below: 

**Procedure:** 
Arm and Takeoff in position mode, after flying for approximately one minute, switched to altitude then stabilized mode proceeds to switch to mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, once completed all waypoint activate RTL.

**Notes:
Please note on altitude mode, when doing pitch and the sticks centered the vehicle would climb altitude. On landing it did a couple  small leaps then landed and disarmed. Other than that good flight no visible observations of any issues.** 

**Log:**  https://review.px4.io/plot_app?log=aff566df-40f7-41fa-92d7-75e9992c9fa3

File: boards/px4/fmu-v4/nuttx-config/include/board.h [hunk 4]
```diff
@@ -269,6 +269,10 @@
 #define GPIO_SPI1_MOSI  (GPIO_SPI1_MOSI_1|GPIO_SPEED_50MHz)
 #define GPIO_SPI1_SCK   (GPIO_SPI1_SCK_1|GPIO_SPEED_50MHz)
 
+#define DMACHAN_SPI1_RX DMAMAP_SPI1_RX_1
+#define DMACHAN_SPI1_TX DMAMAP_SPI1_TX_2
```
Comment (bkueng) [inline boards/px4/fmu-v4/nuttx-config/include/board.h]: This one conflicts with DShot (TIM1_UP). We can use `DMAMAP_SPI1_TX_1` and move the SDIO to stream 6 (`DMAMAP_SDIO_2`).

Comment (dagar) [reply boards/px4/fmu-v4/nuttx-config/include/board.h]: Thanks, I haven't really looked at where things settled after dshot came in. We need to do a pass to get this in place for most of the boards. https://github.com/PX4/Firmware/pull/12436

Comment (dagar) [reply boards/px4/fmu-v4/nuttx-config/include/board.h]: I've started capturing the full mapping in board_dma_map.h (next to board.h).

Comment (bkueng) [reply boards/px4/fmu-v4/nuttx-config/include/board.h]: The current state is here: https://docs.google.com/spreadsheets/d/1mFul5rEZUhfWSB8qSNtGBibBrFA8MPpm0KkxnSyVReI/edit#gid=0. Moving this in-tree is fine with me, it should just be easy to see what are the alternative options.

**Comment by dagar** [comment]
Ok new plan. Let's get this drivers in on the side inactive by default and without the ecl clipping changes. They're still useful for analyzing vibration and we can switch over later once we can better handle clipping and get a lot more testing.

File: src/modules/mavlink/mavlink_messages.cpp [hunk 5]
```diff
@@ -2694,11 +2702,30 @@ class MavlinkStreamEstimatorStatus : public MavlinkStream
...
 
 			// VIBRATION
-			mavlink_vibration_t msg = {};
+			mavlink_vibration_t msg{};
 			msg.time_usec = est.timestamp;
 			msg.vibration_x = est.vibe[0];
 			msg.vibration_y = est.vibe[1];
 			msg.vibration_z = est.vibe[2];
+
+			sensor_accel_status_s acc_status_0;
```
Comment (dagar) [inline src/modules/mavlink/mavlink_messages.cpp]:
```suggestion
			sensor_accel_status_s acc_status_0{};
```

File: src/modules/mavlink/mavlink_messages.cpp [hunk 6]
```diff
@@ -2694,11 +2702,30 @@ class MavlinkStreamEstimatorStatus : public MavlinkStream
...
+
+			if (_sensor_accel_status_0_sub->update(&acc_status_0)) {
+				msg.clipping_0 = acc_status_0.clipping[0] + acc_status_0.clipping[1] + acc_status_0.clipping[2];
+			}
+
+			sensor_accel_status_s acc_status_1;
```
Comment (dagar) [inline src/modules/mavlink/mavlink_messages.cpp]:
```suggestion
			sensor_accel_status_s acc_status_1{};
```

**Comment by dagar** [comment]
Merging this to enable further testing and iteration. Note, these new drivers must be manually started (with the existing mpu6000 stopped).

``` Console
nsh> mpu6000 stop

nsh> icm20602 -R 8 start
icm20602 on SPI bus 1 at 5 (10000 KHz)

nsh> listener sensor_gyro_fifo

TOPIC: sensor_gyro_fifo
 sensor_gyro_fifo_s
        timestamp: 43686752  (0.000431 seconds ago)
        timestamp_sample: 43685585
        device_id: 3671306 (Type: 0x38, SPI:1 (0x05)) 
        dt: 125.0000
        scale: 0.0010
        x: [-27, -21, -33, -28, -25, -33, -16, -15]
        y: [-7, -25, -33, -21, -30, -10, 0, -7]
        z: [-8, -6, -4, -8, 10, -7, 3, 8]
        samples: 8

nsh> listener sensor_accel_status

TOPIC: sensor_accel_status
 sensor_accel_status_s
        timestamp: 50664663  (0.000564 seconds ago)
        error_count: 0
        clipping: [0, 0, 0]
        device_id: 3605770 (Type: 0x37, SPI:1 (0x05)) 
        temperature: 46.9094
        full_scale_range: 156.9064
        measure_rate: 1000
        sample_rate: 4000
        rotation: 8
```

**Comment by mhkabir** [comment]
What is the plan for the ICM support currently in the mpu6000 driver?

**Comment by dagar** [comment]
> What is the plan for the ICM support currently in the mpu6000 driver?

Nothing until these are complete replacements. At that point the mpu6000 itself will be stripped down to be nothing but an optimized mpu6000 driver.

The real blockers for full deployment are:
 - SPI DMA support everywhere (currently there are minor issues on F7 and H7)
 - estimator robustness to IMU clipping

In the testing so far during periods of accelerometer clipping the immediate results are quite bad (enormous velocity estimate), and appear to be worse than the behavior of the current driver where the DLPF is enabled and the clipped accelerometer readings are obscured.

Screen Shot 2019-12-03 at 11 02 35 PM

**Comment by sukshata** [comment]
Hi I am looking for ICM20608-G sensor driver code base migration can anyone help me with the github link to do the same?

**Comment by dagar** [comment]
> Hi I am looking for ICM20608-G sensor driver code base migration can anyone help me with the github link to do the same?

The new icm20608g driver is here. https://github.com/PX4/Firmware/tree/master/src/drivers/imu/invensense/icm20608g

**Comment by sukshata** [comment]
@dagar Thank you so much, I am seeing one issue with mpu9250 on android 8 after the flash device shows up the accelerometer values properly on the app we have but after the device reboot it does not show.

My initial log analysis is Sensor service is not receiving the events back even after sending the flush event and in the HAL i see the events have been received from the kernel.

But HAL is not calling buildmpuevent, can anyone help  me to see the mpu9250  known issues or mpu6250 code base link to see any fixes have gone for the above issue.

Any help to debug this issue would be very helpful, I am new bee for this domain.
Thanks in advance.

**Pull request merged**